src/libcore/slice.rs

```diff
@@ -632,35 +632,17 @@ fn size_from_ptr<T>(_: *const T) -> usize {
 
 
 // Use macros to be generic over const/mut
-//
-// They require non-negative `$by` because otherwise the expression
-// `(ptr as usize + $by)` would interpret `-1` as `usize::MAX` (and
-// thus trigger a panic when overflow checks are on).
-
-// Use this to do `$ptr + $by`, where `$by` is non-negative.
-macro_rules! slice_add_offset {
+macro_rules! slice_offset {
     ($ptr:expr, $by:expr) => {{
         let ptr = $ptr;
         if size_from_ptr(ptr) == 0 {
-            transmute(ptr as usize + $by)
+            transmute((ptr as isize).wrapping_add($by))
```

Comment:
@Gankro I believe it's a `transmute` because a) the macro doesn't have any way to refer to the element type so it can't use an `as` cast, and b) so the same code works with both `*const` and `*mut` pointers. But the `transmute` is constrained to producing the same pointer type as `$ptr` already, due to the use of `ptr.offset($by)` in the other branch.